src/praisonai-agents/praisonaiagents/agent/tool_execution.py

```
@@ -2985,6 +2996,31 @@ def _handle_bridge_tool_call(self, function_name: str, arguments: Dict[str, Any]
+        if not isinstance(tools, (list, tuple)):
+            tools = []
+        for tool in tools:
+            name_attr = getattr(tool, '__name__', None) or getattr(tool, 'name', None)
+            if name_attr == tool_name:
+                explicit = getattr(tool, 'idempotent', None)
+                if isinstance(explicit, bool):
+                    return not explicit
+                break
+        try:
+            from ..escalation.loop_guard import MUTATING_TOOLS
+        except Exception:
```

Ruff (0.16.1)

[warning] Do not catch blind exception: `Exception`

(BLE001)